flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerHeap.java

```diff
@@ -0,0 +1,504 @@
+		final TimerHeapInternalTimer<K, N>[] heap = this.queue;
+		TimerHeapInternalTimer<K, N> currentTimer = heap[idx];
+		int parentIdx = idx >>> 1;
+
+		while (parentIdx > 0 && COMPARATOR.compare(currentTimer, heap[parentIdx]) < 0) {
+			moveElementToIdx(heap[parentIdx], idx);
+			idx = parentIdx;
+			parentIdx = parentIdx >>> 1;
+		}
+
+		moveElementToIdx(currentTimer, idx);
+	}
+
+	private void siftDown(int idx) {
+		final TimerHeapInternalTimer<K, N>[] heap = this.queue;
+		final int heapSize = this.size;
+
+		TimerHeapInternalTimer<K, N> currentTimer = heap[idx];
+		int firstChildIdx = idx << 1;
+		int secondChildIdx = firstChildIdx + 1;
+
+		if (secondChildIdx <= heapSize && COMPARATOR.compare(heap[secondChildIdx], heap[firstChildIdx]) < 0) {
+			firstChildIdx = secondChildIdx;
+		}
+
+		while (firstChildIdx <= heapSize && COMPARATOR.compare(heap[firstChildIdx], currentTimer) < 0) {
+			moveElementToIdx(heap[firstChildIdx], idx);
+			idx = firstChildIdx;
+			firstChildIdx = idx << 1;
+			secondChildIdx = firstChildIdx + 1;
+
+			if (secondChildIdx <= heapSize && COMPARATOR.compare(heap[secondChildIdx], heap[firstChildIdx]) < 0) {
+				firstChildIdx = secondChildIdx;
+			}
+		}
+
+		moveElementToIdx(currentTimer, idx);
+	}
+
+	/**
+	 * Returns an unmodifiable set of all timers in the given key-group.
+	 */
+	Set<InternalTimer<K, N>> getTimersForKeyGroup(@Nonnegative int keyGroupIdx) {
+		return Collections.unmodifiableSet(getDedupMapForKeyGroup(keyGroupIdx).keySet());
+	}
+
+	private HashMap<TimerHeapInternalTimer<K, N>, TimerHeapInternalTimer<K, N>> getDedupMapForKeyGroup(
+		@Nonnegative int keyGroupIdx) {
+		return deduplicationMapsByKeyGroup[globalKeyGroupToLocalIndex(keyGroupIdx)];
+	}
+
+	@VisibleForTesting
+	@SuppressWarnings("unchecked")
+	List<Set<InternalTimer<K, N>>> getTimersByKeyGroup() {
+		List<Set<InternalTimer<K, N>>> result = new ArrayList<>(deduplicationMapsByKeyGroup.length);
+		for (int i = 0; i < deduplicationMapsByKeyGroup.length; ++i) {
+			result.add(i, Collections.unmodifiableSet(deduplicationMapsByKeyGroup[i].keySet()));
+		}
+		return result;
+	}
+
+	private void moveElementToIdx(TimerHeapInternalTimer<K, N> element, int idx) {
+		queue[idx] = element;
+		element.setTimerHeapIndex(idx);
+	}
+
+	private HashMap<TimerHeapInternalTimer<K, N>, TimerHeapInternalTimer<K, N>> getDedupMapForKeyGroup(
+		TimerHeapInternalTimer<?, ?> timer) {
+		int keyGroup = KeyGroupRangeAssignment.assignToKeyGroup(timer.getKey(), totalNumberOfKeyGroups);
+		return getDedupMapForKeyGroup(keyGroup);
+	}
+
+	private int globalKeyGroupToLocalIndex(int keyGroup) {
+		checkArgument(keyGroupRange.contains(keyGroup));
+		return keyGroup - keyGroupRange.getStartKeyGroup();
+	}
+
+	private void checkCapacity(int requested) {
+		int oldArraySize = queue.length;
+
+		if (requested >= oldArraySize) {
+			final int grow = (oldArraySize < 64) ? oldArraySize + 2 : oldArraySize >> 1;
+			int newArraySize = oldArraySize + grow;
```

Comment:
Right, the code was initially only used to handle growth in steps of 1 from the `add` method. We can adjust it a bit so that bulk inserts take less growth steps.